### PR TITLE
Correct issue with parallel_reduce and LaunchBounds value smaller than 32

### LINKED ISSUE #8443: `parallel_reduce` with `LaunchBounds` smaller than 32 returns incorrect results on CUDA backends
The following code fails on CUDA backends:
```C++
template <class T>
void test_func() {
  Kokkos::View<T*> v("View", 32);
  Kokkos::deep_copy(v, T(10));

  // Functor
  auto l = KOKKOS_LAMBDA(const int i, T& ret) {
    ret += v(i);
  };

  Kokkos::parallel_reduce(Kokkos::RangePolicy(0, 1), l, Kokkos::Sum<T>(ret1));
  EXPECT_EQ(ret1, 10); // <= OK

  // The bug is present for all LaunchBounds between 1 and 31.
  Kokkos::parallel_reduce(Kokkos::RangePolicy<Kokkos::LaunchBounds<1>>(0, 1), 
      l, Kokkos::Sum<T>(ret1)); // Problem also appears with Prod but not the other Reducers.
  EXPECT_EQ(ret1, 10); // <= here, ret1 contains a seemingly random value, 110 for LaunchBounds<1>
```

It looks like some variables are not initialised in `parallel_reduce` when `LaunchBounds` is set to a value smaller than 32.

I'm pretty sure that using such low value for `LaunchBounds` won't bring any performance benefit, but I couldn't find anything in the Cuda documentation that would prevent it from working (and the tests I did suggest it should work) so it seems to be a bug inside Kokkos.

### PR DESCRIPTION
**Pull request by PaulGannay**
Create a test to check for issue https://github.com/kokkos/kokkos/issues/8443 and propose a fix.

### CODE REVIEW

File: core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp [hunk 1]
```diff
@@ -243,7 +245,7 @@ struct CudaReductionsFunctor<FunctorType, false, true> {
     if (warp_id == 0) {
       functor.init(&value);
       for (unsigned int i = threadIdx.y * blockDim.x + threadIdx.x;
-           i < blockDim.y * blockDim.x / 32; i += 32)
+           i < num_warps; i += 32)
         functor.join(&value, &shared_team_buffer_element[i]);
       scalar_intra_warp_reduction(functor, value, false, 32,
```
Comment (masterleinad) [inline core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp]: Should't this `32` also be replaced? Did you just keep it because the shared memory buffer would be zero-initialized (neutral element initialized) anyway?

Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp]: It should be fine since we are calling `functor.init(&value)`.

File: core/src/HIP/Kokkos_HIP_ReduceScan.hpp [hunk 2]
```diff
@@ -166,24 +168,24 @@ struct HIPReductionsFunctor<FunctorType, false> {
...
       int const /*shared_elements*/, Scalar* shared_team_buffer_element) {
     int const warp_id = (threadIdx.y * blockDim.x) / HIPTraits::WarpSize;
+    const unsigned int num_threads = blockDim.x * blockDim.y;
     Scalar* const my_shared_team_buffer_element =
         shared_team_buffer_element + threadIdx.y * blockDim.x + threadIdx.x;
     *my_shared_team_buffer_element = value;
     // Warp Level Reduction, ignoring Kokkos vector entries
     scalar_intra_warp_reduction(functor, my_shared_team_buffer_element, skip,
-                                HIPTraits::WarpSize);
+                                Kokkos::min(num_threads, HIPTraits::WarpSize));
```
Comment (masterleinad) [inline core/src/HIP/Kokkos_HIP_ReduceScan.hpp]:
```
In file included from /builds/kokkos/kokkos/core/src/HIP/Kokkos_HIP_Team.hpp:28:
/builds/kokkos/kokkos/core/src/HIP/Kokkos_HIP_ReduceScan.hpp:177:33: error: no matching function for call to 'min'
  177 |                                 Kokkos::min(num_threads, HIPTraits::WarpSize));
      |                                 ^~~~~~~~~~~
/builds/kokkos/kokkos/core/src/Kokkos_MinMax.hpp:72:43: note: candidate template ignored: deduced conflicting types for parameter 'T' ('unsigned int' vs. 'int')
   72 | constexpr KOKKOS_INLINE_FUNCTION const T& min(const T& a, const T& b) {
      |                                           ^
/builds/kokkos/kokkos/core/src/Kokkos_MinMax.hpp:101:36: note: candidate template ignored: could not match 'std::initializer_list<T>' against 'unsigned int'
  101 | KOKKOS_INLINE_FUNCTION constexpr T min(std::initializer_list<T> ilist,
      |                                    ^
/builds/kokkos/kokkos/core/src/Kokkos_MinMax.hpp:89:36: note: candidate function template not viable: requires single argument 'ilist', but 2 arguments were provided
   89 | KOKKOS_INLINE_FUNCTION constexpr T min(std::initializer_list<T> ilist) {
      |                                    ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/builds/kokkos/kokkos/core/src/Kokkos_MinMax.hpp:80:43: note: candidate function template not viable: requires 3 arguments, but 2 were provided
   80 | constexpr KOKKOS_INLINE_FUNCTION const T& min(const T& a, const T& b,
      |                                           ^   ~~~~~~~~~~~~~~~~~~~~~~~
   81 |                                               ComparatorType comp) {
      |                                               ~~~~~~~~~~~~~~~~~~~
```

```suggestion
scalar_intra_warp_reduction(functor, my_shared_team_buffer_element, skip,
                                Kokkos::min<int>(num_threads, HIPTraits::WarpSize));
```
or declare `num_threads` as `int`.

File: core/unit_test/TestReducers.hpp [hunk 3]
```diff
@@ -447,6 +447,27 @@ struct TestReducers {
     }
   }
 
+  // Test that reducers return correct results with LaunchBounds value smaller
+  // that 32.
```
Comment (masterleinad) [inline core/unit_test/TestReducers.hpp]:
```suggestion
  // Test that reducers return correct results with LaunchBounds value smaller
  // than 32.
```

File: core/unit_test/TestReducers.hpp [hunk 4]
```diff
@@ -447,6 +447,27 @@ struct TestReducers {
...
+    // This first reduction is necessary to ensure we trigger the bug #8443
+    Scalar ret;
+    Kokkos::parallel_reduce(Kokkos::RangePolicy(0, 1), l,
+                            Kokkos::Sum<Scalar>(ret));
+    EXPECT_EQ(ret, 10);
+
+    // Test that LaunchBounds<1> works
+    Kokkos::parallel_reduce(Kokkos::RangePolicy<Kokkos::LaunchBounds<1>>(0, 1),
+                            l, Kokkos::Sum<Scalar>(ret));
+    EXPECT_EQ(ret, 10);
```
Comment (masterleinad) [inline core/unit_test/TestReducers.hpp]: I would prefer to also test something where the identity is not given by default/zero initialization, say Kokkos::Prod.

Comment (PaulGannay) [reply core/unit_test/TestReducers.hpp]: Good point, I will test with both.

File: core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp]: It might still be better to adjust 32 for consistency.

Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp]: The implementation of `scalar_intra_warp_reduction` is
```
__device__ static inline void scalar_intra_warp_reduction(
      const FunctorType& functor,
      Scalar value,            // Contribution
      const bool skip_vector,  // Skip threads if Kokkos vector lanes are not
                               // part of the reduction
      const int width,         // How much of the warp participates
      Scalar& result) {
    unsigned mask =
        width == 32
            ? 0xffffffff
            : ((1 << width) - 1)
                  << ((threadIdx.y * blockDim.x + threadIdx.x) / width) * width;
    for (int delta = skip_vector ? blockDim.x : 1; delta < width; delta *= 2) {
      Scalar tmp = Kokkos::shfl_down(value, delta, width, mask);
      functor.join(&value, &tmp);
    }
    
    Impl::in_place_shfl(result, value, 0, width, mask);
  }
```
and the loop in the middle might produce garbage values without the mask if we are trying to access values from threads that don't exist (because the warp isn't full).

**Review by crtrott — COMMENTED** [review]
How about we just static assert the LaunchBounds being a multiple of 32 for CUDA/HIP/SYCL?

**Comment by crtrott** [comment]
I am not sure that this needs fixing so much as simply testing for valid values. Thoughts?

**Comment by PaulGannay** [comment]
Value that are not multiple of 32 are valid in CUDA, and I believe in HIP (but I need to do more rigorous testing) so there is no reason to forbid it.

**Comment by crtrott** [comment]
Yeah they are valid, but they are also useless, because registers are allocated in banks of 32 or so.

**Comment by PaulGannay** [comment]
It is useless, but users may stumble upon it if they use heuristics to choose their launch bounds.

File: core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp (hunk 1 above)
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp]: Yeah I think this is problematic not sure our tests will catch it but I am not convinced this is correct.

**Comment by masterleinad** [comment]
> How about we just static assert the LaunchBounds being a multiple of 32 for CUDA/HIP/SYCL?

I'd prefer that.

**Comment by crtrott** [comment]
Looks like HIP testing is catching this as wrong - that is more likely btw since on HIP you do have fixed lockstep warps. Weird thing is serial test for reducer_int also failed. I am wondering if we missing a "ExecutionSpace" argument somewhere in that test, so that it runs HIP twice.

File: core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp (hunk 1 above)
Comment (PaulGannay) [reply core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp]: > Should't this `32` also be replaced? Did you just keep it because the shared memory buffer would be zero-initialized (neutral element initialized) anyway?

~~It seems ok because to get here, either:~~
 - ~~we have less than 32 threads, which means that `num_warps == 0`, `i` can't be strictly smaller than 0 so we don't reach line 250,~~
 - ~~we have 32 threads or more, since only the first warp is accessing this section, all 32 threads should be active.~~
=> all of the above is wrong, I was tricked by the absence of bracket around the loop body
@masterleinad is right, the 32 should be replaced too.

Comment (cwpearson) [reply core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp]: As this is confusing enough to warrant a PR discussion, let's get some comments in the code too.

**Review by cwpearson — CHANGES_REQUESTED** [review]

File: core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp (hunk 1 above)
Comment (PaulGannay) [reply core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp]: > As this is confusing enough to warrant a PR discussion, let's get some comments in the code too.

I was wrong in my explanation but you are right, it deserve a comment, thank you.

**Review by crtrott — APPROVED** [review]

**Comment by crtrott** [comment]
Retest this please.

**Review by cwpearson — APPROVED** [review]
I'm good with it.

**Comment by crtrott** [comment]
Retest this please.
